### PR TITLE
[SR GeekCamp]Support alter table rename table clause in new Parser and new Analyzer

### LINKED ISSUE #4042: New Parser and Analyzer support table and partition related statements【Easy】
# Requirement Description
The new Parser and the new Analyzer support table and partition related statements

# Mentor
@HangyuanLiu 

# Registration procedure

**Step 1.** Add a comment when you sign up. The comment can be a brief self-introduction and reason for signing up, including what you do, what you're good at, what you're interested in and your technology stack, and how familiar you are with StarRocks. Example:

> I pick! My name is Johnny. I am from StarRocks and I have been working on database kernel development for 1 year. I am interested in OLAP, good at C++ programming, and have been using StarRocks for 6 months. I chose to solve this issue because I am familiar with execution engine.

**Step 2.** On your WeChat, add the StarRocks community assistant "**_StarRocks-1_**" as a WeChat friend and leave your GitHub ID. The assistant will invite you to a WeChat group based on the issue you picked.

**Step 3.** The mentor will communicate with all the participants who picked this issue and assign the task to the best match. 

# Before you start
1. Read the StarRocks [GitHub workflow](https://github.com/StarRocks/community/blob/main/Contributors/guide/workflow.md) carefully, which provides the configurations of BE and FE and describes how to submit a pull request.
2. To learn more about StarRocks, read [StarRocks documentation](https://github.com/StarRocks/docs/blob/main/TOC.md).

# Pull request requirements

1. After you complete code development, upload the code to the StarRocks code repo as a pull request (PR).

  - PR title format: **[SR GeekCamp] + PR title**
  - Labels ：[Geek Camp 2022/feature]
2. After you submit your PR, comment on the status of your PR submission as requested under the Pinned Issue [[StarRocks geek Camp 2022] ](https://github.com/StarRocks/starrocks/issues/4008)page.

  - Comment format: **"I have finished Issue #" + PR link id**

### PR DESCRIPTION
**Pull request by zhuxt2015**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：
Fixes #4042

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 48 / 64 (75.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|4|50.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableClauseAnalyzer.java|10|14|71.43%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterStmtAnalyzer.java|20|27|74.07%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|8|9|88.89%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/TableRenameClause.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterTableStmt.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 49 / 62 (79.03%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|4|50.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterStmtAnalyzer.java|19|25|76.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableClauseAnalyzer.java|12|14|85.71%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|8|9|88.89%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/TableRenameClause.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterTableStmt.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 49 / 60 (81.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterStmtAnalyzer.java|19|25|76.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableClauseAnalyzer.java|12|14|85.71%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|8|9|88.89%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/TableRenameClause.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterTableStmt.java|1|1|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableClauseAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import com.starrocks.analysis.AlterClause;
+import com.starrocks.analysis.TableRenameClause;
+import com.starrocks.common.AnalysisException;
+import com.starrocks.common.ErrorCode;
+import com.starrocks.common.ErrorReport;
+import com.starrocks.common.FeNameFormat;
+import com.starrocks.qe.ConnectContext;
+import com.starrocks.sql.ast.AstVisitor;
+
+public class AlterTableClauseAnalyzer {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableClauseAnalyzer.java:14]: Combine AlterStmtAnalyzerVisitor and AlterTableClauseAnalyzer into an AlterTableStatementAnalyzer.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterStmtAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            try {
+                CatalogUtils.checkOlapTableHasStarOSPartition(tbl.getDb(), tbl.getTbl());
+            } catch (AnalysisException e) {
+                ErrorReport.reportSemanticException(ErrorCode.ERR_NO_TABLES_USED);
+            }
+            List<AlterClause> alterClauseList = statement.getOps();
+            if (alterClauseList == null || alterClauseList.isEmpty()) {
+                ErrorReport.reportSemanticException(ErrorCode.ERR_NO_ALTER_OPERATION);
+            }
+            com.starrocks.analysis.Analyzer analyzer = new Analyzer(context.getCatalog(), context);
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterStmtAnalyzer.java:43]: can't use old analyzer, alter table clause not implement can't reach here

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 47 / 55 (85.45%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|29|34|85.29%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|8|9|88.89%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/TableRenameClause.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterTableStmt.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 39 / 50 (78.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|6|9|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|23|29|79.31%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/TableRenameClause.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterTableStmt.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 39 / 49 (79.59%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|6|9|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|23|28|82.14%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/TableRenameClause.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterTableStmt.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 41 / 49 (83.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|23|28|82.14%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|8|9|88.89%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/TableRenameClause.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterTableStmt.java|1|1|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import com.starrocks.common.ErrorReport;
+import com.starrocks.common.FeNameFormat;
+import com.starrocks.qe.ConnectContext;
+import com.starrocks.sql.StatementPlanner;
+import com.starrocks.sql.ast.AstVisitor;
+import com.starrocks.sql.common.MetaUtils;
+
+import java.util.List;
+
+public class AlterTableStatementAnalyzer {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java:21]: Miss logic of PrivilegeChecker

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        List<AlterClause> alterClauseList = statement.getOps();
+        if (alterClauseList == null || alterClauseList.isEmpty()) {
+            ErrorReport.reportSemanticException(ErrorCode.ERR_NO_ALTER_OPERATION);
+        }
+        AlterTableClauseAnalyzerVisitor alterTableClauseAnalyzerVisitor = new AlterTableClauseAnalyzerVisitor();
+        for (AlterClause alterClause : alterClauseList) {
+            if (StatementPlanner.isNewAlterTableClause(alterClause)) {
+                alterTableClauseAnalyzerVisitor.analyze(alterClause, context);
+            } else {
+                throw new SemanticException(alterClause.getOpType().name() + " clause not support new Analyzer");
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java:39]:
```suggestion
        throw new StarRocksPlannerException(alterClause.getOpType().name() + " clause not support new Analyzer", ErrorType.INTERNAL_ERROR);
```

File: fe/fe-core/src/test/java/com/starrocks/alter/AlterTest.java [hunk 5]
```diff
@@ -159,7 +157,7 @@ private static void createTable(String sql) throws Exception {
     }
 
     private static void alterTable(String sql, boolean expectedException) throws Exception {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/test/java/com/starrocks/alter/AlterTest.java:159]: Don't modify UtFrameUtils.parseStmtWithNewParser, you can add an alterTableWithNewParser function

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 45 / 54 (83.33%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|4|5|80.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|23|28|82.14%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|8|9|88.89%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/TableRenameClause.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterTableStmt.java|1|1|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeChecker.java [hunk 6]
```diff
@@ -32,6 +33,16 @@ public void check(StatementBase statement, ConnectContext session) {
             visit(statement, session);
         }
 
+        @Override
+        public Void visitAlterTableStatement(AlterTableStmt statement, ConnectContext session) {
+            String dbName = statement.getTbl().getDb();
+            String tableName = statement.getTbl().getTbl();
+            if (!Catalog.getCurrentCatalog().getAuth().checkTblPriv(session, dbName, tableName, PrivPredicate.ALTER)) {
+                ErrorReport.reportSemanticException(ErrorCode.ERR_SPECIFIC_ACCESS_DENIED_ERROR, "DROP");
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PrivilegeChecker.java:41]: code copy error？

**Comment by HangyuanLiu** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 42 / 54 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|6|9|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|22|28|78.57%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|4|5|80.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/TableRenameClause.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterTableStmt.java|1|1|100.00%|

**Review by HangyuanLiu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 42 / 54 (77.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|6|9|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|22|28|78.57%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|4|5|80.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/TableRenameClause.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/AlterTableStmt.java|1|1|100.00%|

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**